### PR TITLE
Replace third party slack library with official one

### LINKED ISSUE #743: Memory leak when using slack connector
I've noticed in a couple of different deployments of opsdroid that there seems to be a memory leak. In my work deployment it seems to happen at the same time every day so it is likely related to the cron timing. It looks like the memory just keeps filling until the OOM killer ends the process and it restarts again. I'll have a look through the skills I have and see if anything lines up with the timings.

image

image

### LINKED ISSUE #845: Slack Connector: Websockets exception raised - task pending
# Description

I tested opsdroid on my kali machine for the first time with the slack, shell and websockets connector set up. I stopped opsdroid and an exception happened.

By the looks of the traceback, it seems that it happened because websockets `wait_for_connection_lost` before stopping, at the same time we are cancelling every pending tasking.

Perhaps we should add a delay before cancelling pending tasks, or prioritize the disconnect methods on every connector and remove the pending tasks.

## Steps to Reproduce
Install Kali, start opsdroid with slack, shell and websockets, `ctrl+c` Exception happens

## Expected Functionality
Opsdroid should exit gracefully

## Experienced Functionality

```shell
root@kali:~# opsdroid
INFO opsdroid: ========================================
INFO opsdroid: Started application
INFO opsdroid: ========================================
INFO opsdroid: You can customise your opsdroid by modifying your configuration.yaml
INFO opsdroid: Read more at: http://opsdroid.readthedocs.io/#configuration
INFO opsdroid: Watch the Get Started Videos at: http://bit.ly/2fnC0Fh
INFO opsdroid: Install Opsdroid Desktop at: 
https://github.com/opsdroid/opsdroid-desktop/releases
INFO opsdroid: ========================================
WARNING opsdroid.loader: No databases in configuration.This will cause skills which store things in memory to lose data when opsdroid is restarted.
WARNING opsdroid.core: Function based skills are deprecated and will be removed in a future release. Please use class-based skills instead.
WARNING opsdroid.core: Function based skills are deprecated and will be removed in a future release. Please use class-based skills instead.
WARNING opsdroid.core: Function based skills are deprecated and will be removed in a future release. Please use class-based skills instead.
WARNING opsdroid.core: Function based skills are deprecated and will be removed in a future release. Please use class-based skills instead.
WARNING opsdroid.core: Function based skills are deprecated and will be removed in a future release. Please use class-based skills instead.
WARNING opsdroid.core: Function based skills are deprecated and will be removed in a future release. Please use class-based skills instead.
INFO opsdroid.connector.slack: Connecting to Slack
INFO opsdroid.connector.slack: Connected successfully
INFO opsdroid.core: Opsdroid is now running, press ctrl+c to exit.
INFO opsdroid.web: Started web server on http://127.0.0.1:8080
opsdroid> INFO opsdroid.core: Received stop signal, exiting.
INFO opsdroid.core: Removing skills...
INFO opsdroid.core: Removed dance
INFO opsdroid.core: Removed hello
INFO opsdroid.core: Removed seen
INFO opsdroid.core: Stopping connector websocket...
INFO opsdroid.core: Stopped connector websocket
INFO opsdroid.core: Stopping connector slack...
INFO opsdroid.core: Stopped connector slack^C
INFO opsdroid.core: Stopping web server...
INFO opsdroid.core: Stopped web server
INFO opsdroid.core: Stopping cron...
INFO opsdroid.core: Stopped cron
INFO opsdroid.core: Stopping pending tasks...
INFO opsdroid.core: Stopped pending tasks
INFO opsdroid.core: Bye!
INFO opsdroid.core: Exiting application with return code 0
ERROR opsdroid.core: Caught exception
ERROR opsdroid.core: {'message': 'Task was destroyed but it is pending!', 'task': <Task pending coro=<WebSocketCommonProtocol.close_connection() running at /usr/local/lib/python3.7/dist-packages/websockets/protocol.py:1029> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7fb6d9703948>()]> cb=[gather.<locals>._done_callback() at /usr/lib/python3.7/asyncio/tasks.py:664]>}
Exception ignored in: <generator object WebSocketCommonProtocol.close_connection at 0x7fb6d9750c00>
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/websockets/protocol.py", line 1056, in close_connection
    if (yield from self.wait_for_connection_lost()):
  File "/usr/local/lib/python3.7/dist-packages/websockets/protocol.py", line 1080, in wait_for_connection_lost
    loop=self.loop,
  File "/usr/lib/python3.7/asyncio/tasks.py", line 400, in wait_for
    timeout_handle = loop.call_later(timeout, _release_waiter, waiter)
  File "/usr/lib/python3.7/asyncio/base_events.py", line 657, in call_later
    context=context)
  File "/usr/lib/python3.7/asyncio/base_events.py", line 667, in call_at
    self._check_closed()
  File "/usr/lib/python3.7/asyncio/base_events.py", line 480, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
```

## Versions
- **Opsdroid version:** Latest
- **Python version:** 3.7.2
- **OS/Docker version:** Kali Linux Rolling

## Configuration File
Please include your version of the configuration file bellow.

```yaml
##                      _           _     _
##   ___  _ __  ___  __| |_ __ ___ (_) __| |
##  / _ \| '_ \/ __|/ _` | '__/ _ \| |/ _` |
## | (_) | |_) \__ \ (_| | | | (_) | | (_| |
##  \___/| .__/|___/\__,_|_|  \___/|_|\__,_|
##       |_|
##                   __ _
##   ___ ___  _ __  / _(_) __ _
##  / __/ _ \| '_ \| |_| |/ _` |
## | (_| (_) | | | |  _| | (_| |
##  \___\___/|_| |_|_| |_|\__, |
##                        |___/
##
## A default config file to use with opsdroid

## Set the logging level
logging:
#   level: info
#   path: opsdroid.log
   console: true

## Set the location for opsdroid to install modules
# module-path: "."

## Show welcome message
welcome-message: true

## Configure the web server
# web:
#   host: '127.0.0.1'
#   port: 8080
#   ssl:
#     cert: /path/to/cert.pem
#     key: /path/to/key.pem

## Parsers
# parsers:
#
#  ## Dialogflow (http://opsdroid.readthedocs.io/en/stable/matchers/dialogflow)
#  - name: dialogflow
#    access-token: "exampleaccesstoken123"
#    min-score: 0.6
#
#  ## Rasanlu (http://opsdroid.readthedocs.io/en/stable/matchers/rasanlu)
#  - name: rasanlu
#    url: http://localhost:5000
#    project: opsdroid
#    token: 85769fjoso084jd
#    min-score: 0.8
#
#  ## Recast.ai (http://opsdroid.readthedocs.io/en/stable/matchers/recast.ai)
#  - name: recastai
#    access-token: 85769fjoso084jd
#    min-score: 0.8
#
#  ## Regex (http://opsdroid.readthedocs.io/en/stable/matchers/regex)
#  - name: regex
#    enabled: true
#
#  ## Wit.ai (http://opsdroid.readthedocs.io/en/stable/matchers/wit.ai)
#  - name: witai
#    access-token: XJF475SKGITJ98KHFO
#    min-score: 0.6
#
#  ## Luis.ai (http://opsdroid.readthedocs.io/en/stable/matchers/luis.ai)
#   - name: luisai
#     appid: "<application-id>"
#     appkey: "<subscription-key>"
#     verbose: True
#     min-score: 0.6
#
#  ## Crontab (http://opsdroid.readthedocs.io/en/stable/matchers/crontab)
#  - name: crontab
#    enabled: true
#

## Connector modules
connectors:

  - name: websocket
    # optional
    bot-name: "mybot" # default "opsdroid"
    max-connections: 10 # default is 10 users can be connected at once
    connection-timeout: 10 # default 10 seconds before requested socket times out

### Uncomment the connector(s) that you wish opsdroid to work on
#
#  - name: shell
#    # optional
#    bot-name: "mybot" # default "opsdroid"
#
#  ## Ciscospark (https://github.com/opsdroid/connector-ciscospark)
#  - name: ciscospark
#    # required
#    webhook-url: http(s)://<host>:<port>  # Url for Spark to connect to your bot
#    access-token: <your bot access token>  # Your access token
#
#  ## Facebook (https://github.com/opsdroid/connector-facebook)
#  - name: facebook
#    # required
#    verify-token: aabbccddee
#    page-access-token: aabbccddee112233445566
#    # optional
#    bot-name: "mybot" # default "opsdroid"
#
#  ## Github (https://github.com/opsdroid/connector-github)
#  - name: github
#    # required
#    github-token: aaabbbcccdddeee111222333444
#
#  ## Matrix (https://github.com/opsdroid/connector-matrix)
#  - name: matrix
#    # Required
#    mxid: "@username:matrix.org"
#    password: "mypassword"
#    # Name of a single room to connect to
#    room: "#matrix:matrix.org"
#    # Alternatively, a dictionary of multiple rooms
#    # One of these should be named 'main'
#    rooms:
#      'main': '#matrix:matrix.org'
#      'other': '#riot:matrix.org'
#    # Optional
#    homeserver: "https://matrix.org"
#    nick: "Botty McBotface"  # The nick will be set on startup
#
#  ## Shell (https://github.com/opsdroid/connector-shell)
  - name: shell
#    # optional
#    bot-name: "mybot" # default "opsdroid"
#
#  ## Slack (https://github.com/opsdroid/connector-slack)
  - name: slack
    api-token: $SLACKAPI
#    # required
#    api-token: "zyxw-abdcefghi-12345"
#    # optional
#    bot-name: "mybot" # default "opsdroid"
#    default-room: "#random" # default "#general"
#    icon-emoji: ":smile:" # default ":robot_face:"
#
#  ## Telegram (https://github.com/opsdroid/connector-telegram)
#  - name: telegram
#    # required
#    token: "123456789:ABCDEFGHIJKLMNOPQRSTUVWXYZ-ZYXWVUT"  # Telegraf bot token
#    # optional
#    update_interval: 0.5  # Interval between checking for messages
#    default_user: user1  # Default user to send messages to (overrides default room in connector)
#    whitelisted_users:  # List of users who can speak to the bot, if not set anyone can speak
#      - user1
#      - user2
#
#  ## Twitter (https://github.com/opsdroid/connector-twitter)
#  - name: twitter
#    # required
#    consumer_key: "zyxw-abdcefghi-12345"
#    consumer_secret: "zyxw-abdcefghi-12345-zyxw-abdcefghi-12345"
#    oauth_token: "zyxw-abdcefghi-12345-zyxw-abdcefghi-12345"
#    oauth_token_secret: "zyxw-abdcefghi-12345-zyxw-abdcefghi-12345"
#    # optional
#    enable_dms: true  # Should the bot respond to Direct Messages
#    enable_tweets: true  # Should the bot respond to tweets
#
#  ## Websocket (https://github.com/opsdroid/connector-websocket)
#  - name: websocket
#    # optional
#    bot-name: "mybot" # default "opsdroid"
#    max-connections: 10 # default is 10 users can be connected at once
#    connection-timeout: 10 # default 10 seconds before requested socket times out
#


## Database modules (optional)

# databases:
#
#  ## Mongo (https://github.com/opsdroid/database-mongo)
#  - name: mongo
#    host:       "my host"     # (optional) default "localhost"
#    port:       "12345"       # (optional) default "27017"
#    database:   "mydatabase"  # (optional) default "opsdroid"
#
#  ## Redis (https://github.com/opsdroid/database-redis)
#  - name: redis
#    host:       "my host"     # (optional) default "localhost"
#    port:       "12345"       # (optional) default "6379"
#    database:   "7"           # (optional) default "0"
#
#  ## Sqlite (https://github.com/opsdroid/database-sqlite)
#  - name: sqlite
#    file: "my_file.db"  # (optional) default "~/.opsdroid/sqlite.db"
#    table: "my_table"  # (optional) default "opsdroid"
#


## Skill modules
skills:

  ## Dance (https://github.com/opsdroid/skill-dance)
  - name: dance

  ## Hello (https://github.com/opsdroid/skill-hello)
  - name: hello

  ## Loudnoises (https://github.com/opsdroid/skill-loudnoises)
  - name: loudnoises

  ## Seen (https://github.com/opsdroid/skill-seen)
  - name: seen

  # Configurations for other skills uncomment desired skill to activate it.
#
#  ## Cloudhealth (https://github.com/opsdroid/skill-cloudhealth)
#  - name: aws-billing
#    # Required
#    chapi-key: ABCDEF123456789  # Cloud Health API key for billing alerts
#    # Optional
#    room: "#monitoring"  # Room to send alert to
#    daily-billing-alerts: true  # Announce the previous day's bill each morning
#    monthly-billing-alerts: true  # Announce the previous month's bill each month
#
#  ## Dance (https://github.com/opsdroid/skill-dance)
#  - name: dance
#
#  ## Devtools (https://github.com/opsdroid/skill-devtools)
#  - name: devtools
#
#  ## Dialogflow (https://github.com/opsdroid/skill-dialogflow)
#  - name: dialogflow
#    include:
#      - smalltalk
#    exclude:
#      - smalltalk.agent
#
#  ## Food (https://github.com/opsdroid/skill-food)
#  - name: food
#    api-key: "myapikeyfromfood2fork"  # Required
#
#  ## Github (https://github.com/opsdroid/skill-github)
#  - name: github
#
#  ## Google it (https://github.com/opsdroid/skill-google-it)
#  - name: google-it
#    # Use Google search engine (Default)
#    engine-url: https://www.google.co.uk/
#    query-arg: search?q=
#    # Other search engines that can be used (keep only one uncommented at a time)
#    # Use Bing search engine
#  # engine-url: https://www.bing.com/
#  # query-arg: search?=
#    # Use DuckDuckGo search engine
#  # engine-url: https://duckduckgo.com/
#  # query-arg: ?q=
#    # Use Yahoo search engine
#  # engine-url: http://search.yahoo.com/
#  # query-arg: search?p=
#    # Use Aol search engine
#  # engine-url: https://search.aol.co.uk/aol/
#  # query-arg: search?query=
#    # Use Ask search engine
#  # engine-url: https://uk.ask.com/
#  # query-arg: web?q=
#    # Use Wolframalpha search engine
#  # engine-url: https://www.wolframalpha.com/input/
#  # query-arg: ?i=
#
#  ## Grafana (https://github.com/opsdroid/skill-grafana)
#  - name: grafana
#    room: "#monitoring"  # (Optional) room to send alert to
#
#  ## Hacktoberfest (https://github.com/opsdroid/skill-hacktoberfest)
#  - name: hacktoberfest
#
#  ## Hello (https://github.com/opsdroid/skill-hello)
#  - name: hello
#
#  ## Help (https://github.com/opsdroid/skill-help)
#  skill_name <required_parameter> [optional_parameter] - Description of the skill
#
#  ## Homeassistant (https://github.com/opsdroid/skill-homeassistant)
#  - name: homeassistant
#    room: "#homeassistant"  # (Optional) room to send notifications to
#
#  ## Iss (https://github.com/opsdroid/skill-iss)
#  - name: iss
#    # Required
#    api-key: "mygooglemapsapikey"
#    # Optional
#    zoom: "5"
#    map-size: "1024x768"
#    map-type: "hybrid" # hybrid, satellite or roadmap
#
#  ## Loudnoises (https://github.com/opsdroid/skill-loudnoises)
#  - name: loudnoises
#
#  ## Magpi (https://github.com/opsdroid/skill-magpi)
#  - name: magpi
#    room: "#raspberrypi"  # (Optional) room to send notifications to
#
#  ## Please stand by (https://github.com/opsdroid/skill-please-stand-by)
#  - name: please-stand-by
#
#  ## Random (https://github.com/opsdroid/skill-random)
#  - name: random
#
#  ## Seen (https://github.com/opsdroid/skill-seen)
#  - name: seen
#
#  ## Speakingclock (https://github.com/opsdroid/skill-speakingclock)
#  - name: speakingclock
#
#  ## Travis (https://github.com/opsdroid/skill-travis)
#  - name: travis
#    room: "#monitoring"  # (Optional) room to send alert to
#    travis_endpoint: "org"  # (Optional) endpoint for travis, change to "com" if using enterprise Travis CI
#
#  ## Vault (https://github.com/opsdroid/skill-vault)
#  - name: vault
#    # Required
#    vault-url: https://vault.example.com:8443
#    vault-token: aabbccddee1122334455
#    # Optional
#    announce-on-seal: false  # Announce the vault status in the default room on seal
#    announce-sealed: true  # Announce the vault is sealed hourly
#    announce-unsealed: false  # Announce the vault is unsealed hourly
#
#  ## Weather (https://github.com/opsdroid/skill-weather)
#  - name: weather
#    # Required
#    city: London,UK    # For accuracy use {city},{country code}
#    unit: metric       # Choose metric/imperial
#    api-key: 6fut9e098d8g90g
#
#  ## Word of the day (https://github.com/opsdroid/skill-word-of-the-day)
#  - name: word-of-the-day
#
#  ## Words (https://github.com/opsdroid/skill-words)
#  - name: words
#
```

### LINKED ISSUE #1000: Slack integration uses old API/Token method
# Description
Opsdroid's Slack integration wants to operate off of a Webhook or API Token, which seems to be a deprecating (not yet deprecated) method of integrating with Slack.  See attached screenshot of the warning message they display when trying to do a 'Webhook' or generic 'Bot' integration in an Enterprise Slack instance.  

## Steps to Reproduce
Try to set up a 'Webhook' or 'API/Bot' integration in the usual manner: 
Menu -> Customize Slack -> Configure Apps -> Custom Integrations

Observe Slack's notice that "Custom integrations will be deprecated and possibly removed in the future.  Learn more about replacing them with apps that have more features and use the latest APIs" 

## Expected Functionality
Ideally, Opsdroid would be able to integrate with Slack using their shiny Apps/not-Webhook API, and I don't have to go through a Security Team review because of the deprecation warning :-) 

## Experienced Functionality
Explain what happened instead(Please include the debug log).

## Versions
- **Opsdroid version:** latest
- **Python version:** 3.2
- **OS/Docker version:** latest

## Configuration File
n/a

## Additional Details

[Custom Integration warning](https://i.imgur.com/fWWFXHw.png)

### LINKED ISSUE #1036: Migrate to slackapi/python-slackclient
# Description
With v2, the [python-slackclient library](https://github.com/slackapi/python-slackclient) provided by Slack now supports asyncio natively (https://github.com/slackapi/python-slackclient/blob/add2757ce7ce688ba45355079bafa911a9b61ea5/README.md#async-usage). 

I wanted to mention this in case there was any benefit to migrating to using the Slack-provided library. It may be easier to support new Slack features faster with this library, as well as add support for the RTM API.

### PR DESCRIPTION
**Pull request by jacobtomlinson**
# Description

Swapped out the third party slack library we were using for the [official one](https://github.com/slackapi/python-slackclient) now that it supports async. Users shouldn't notice a difference as these are under the hood changes.

Closes #1036 

**Also**
Closes #845 as it is likely no longer valid for this library
Closes #743 as the new library shouldn't have the memory leak (though it may have its own issues)
Closes #1000 as documentation has been updated with bot user instructions

This does not add support for the webhook based events API as the official library [does not support it](https://github.com/slackapi/python-slackclient#basic-usage-of-the-rtm-client). There is a [second official library](https://github.com/slackapi/python-slack-events-api) which does, however it assumes you are using Flask for your web server which we are not. We should make a separate issue for adding this support ourselves.

### CODE REVIEW

**Comment by jacobtomlinson** [comment]
Also the official library handles a bunch of the websocket stuff itself, so notice this PR deletes more code that it adds 🎉.

File: opsdroid/connector/slack/__init__.py [hunk 1]
```diff
@@ -70,85 +63,65 @@ def __init__(self, config, opsdroid=None):
...
-            if delay is not None:
-                await asyncio.sleep(delay)
-            await self.connect()
-        finally:
-            self.reconnecting = False
-
     async def disconnect(self):
         """Disconnect from Slack."""
-        await self.slacker.close()
+        self.listening = False
```
Comment (Cadair) [inline opsdroid/connector/slack/__init__.py]: Shouldn't these two lines really be the other way around :D

File: opsdroid/connector/slack/__init__.py [hunk 2]
```diff
@@ -70,85 +63,65 @@ def __init__(self, config, opsdroid=None):
...
-                    message["text"],
-                    user_info["name"],
-                    message["channel"],
-                    self,
-                    raw_event=message,
-                )
+        message = payload["data"]
+
+        # Ignore own messages
+        if "bot_id" in message and message["username"] == self.bot_name:
```
Comment (Cadair) [inline opsdroid/connector/slack/__init__.py]: Wouldn't it be more robust to actually check against our bot id? I think you can get that from the auth_info call?

File: opsdroid/connector/slack/__init__.py [hunk 3]
```diff
@@ -70,85 +63,65 @@ def __init__(self, config, opsdroid=None):
...
+        # Lookup username
+        _LOGGER.debug("Looking up sender username")
+        try:
+            user_info = await self.lookup_username(message["user"])
+        except ValueError:
+            return
+
+        # Replace usernames in the message
+        _LOGGER.debug("Replacing userids in message with usernames")
+        message["text"] = await self.replace_usernames(message["text"])
```
Comment (Cadair) [inline opsdroid/connector/slack/__init__.py]: oh this is going to conflict with #1071 :rocket:

File: opsdroid/connector/slack/__init__.py (hunk 1 above)
Comment (jacobtomlinson) [reply opsdroid/connector/slack/__init__.py]: Probably :)

File: opsdroid/connector/slack/__init__.py (hunk 2 above)
Comment (jacobtomlinson) [reply opsdroid/connector/slack/__init__.py]: Yep! Thanks for the pointer.

File: opsdroid/connector/slack/__init__.py (hunk 3 above)
Comment (jacobtomlinson) [reply opsdroid/connector/slack/__init__.py]: Yeah sorry there may be a few breakages with that PR.

Comment (Cadair) [reply opsdroid/connector/slack/__init__.py]: I am sure it wont be that hard to unravel.

File: opsdroid/connector/slack/__init__.py (hunk 2 above)
Comment (jacobtomlinson) [reply opsdroid/connector/slack/__init__.py]: Weirdly the `user_id` that comes back in `auth.text` does not match the `bot_id` that comes with the bot message. Seems there are a few questions about this on Stack Overflow and its not currently possible for a bot to work out its own `bot_id`. The only time it gets to see it is during the OAuth flow, but for opsdroid its normal to just grab the bot token from the console and skip the OAuth flow all together.

Comment (Cadair) [reply opsdroid/connector/slack/__init__.py]: Ok, it's not quite as simple as I thought, it involves two different API calls but you can do it like this:
```python
In [27]: c = slack.WebClient("xoxb-...-...")                                                                                                                                                                                                                                     

In [28]: user_id = c.api_call("auth.test")['user_id']                                                                                                                                                                                                                                                          

In [29]: bot_id = c.users_info(user=user_id)['user']['profile']['bot_id']                                                                                                                                                                                                                                      

In [30]: bot_id                                                                                                                                                                                                                                                                                                
Out[30]: 'B9SBZ44BE'
```

Comment (jacobtomlinson) [reply opsdroid/connector/slack/__init__.py]: Weird I tried users info and was getting an empty response. Let me try again and make sure I'm not being an idiot.

Comment (jacobtomlinson) [reply opsdroid/connector/slack/__init__.py]: My bad, it's working now. I've decided to use the underlying `api_call` method consistently. We had issues in the past with libraries getting out of sync with the API and it's also easier for mocking. But I was making a post rather than a get (which I guess would've been avoided if I used the convenience method).

**Comment by jacobtomlinson** [comment]
This is good to merge on my end.
